rstudio/rstudio
# Add support for cgroups v2 in the environment pane

### Intent

I was working on an unrelated cgroups v2 project and noticed that we don't yet support it. So: this PR adds detection of memory limits and usage when the Linux OS uses cgroups v2. This is the default in Ubuntu 22.04 and Debian 11 and most other distros are expected to switch in the near future.

This topic is tracked in #10914 but this PR does not address the monitor-specific issues raised in that thread.

### Approach

The commit is structured to make minimal behaviour changes to the existing `MemoryProvider` design, but for convenience I've adopted `LONG_MAX` as a sigil value because cgroups v1 and v2 use different representations. This makes some of the existing code simpler.

### Automated Tests

There are existing unit tests; you can run them in the presence of a v2 cgroup by doing something like the following:

```
$ systemd-run --user -p MemoryHigh=100M --pipe --wait build/src/cpp/core/rstudio-core-tests Resources
$ systemd-run --user -p MemoryMax=100M --pipe --wait build/src/cpp/core/rstudio-core-tests Resources
```

### QA Notes

We need to verify that this change (1) works with cgroups v2; but (2) doesn't break cgroups v1. @colearendt and I have checked both of these cases already.

### Checklist
